### ISSUE TITLE
Clarification on Performance and Issues with Sa2VA Models on MeVIS Benchmark

### ISSUE THREAD

**Issue by kumuji** [magic-research/Sa2VA#3]
Hi @lxtGH, @HarborYuan, @zhang-tao-whu,  

Thank you for the awesome work and for sharing Sa2VA.

I am currently trying to reproduce the performance on the MeVIS benchmark using the provided weights and code. While doing so, I slightly modified the demo code to save the output masks, instead of visualizing it. You can find the updated code in my fork:  

[Forked Repository](https://github.com/kumuji/Sa2VA)  
[Modified File](https://github.com/kumuji/Sa2VA/blob/main/demo/demo.py)  

### Observations  
It seems like the setup works as intended, but the performance is lower than expected. Here's a summary of my results using the [Sa2VA-8B](https://huggingface.co/ByteDance/Sa2VA-8B) model:  

1. **Original Sampling Method (First Frames to LLM):**  
   - **`val_u` set**: `J: 54.63`, `F: 63.38`, `J&F: 59.01`  
   - **`val` set**:  `J: 48.15 F: 55.03 J&F: 51.59`  

2. **Uniform Sampling of frames (as it makes more sense for MeVIS):**  
   - **`val_u` set**: `J: 55.88`, `F: 64.26`, `J&F: 60.07`  
   - **`val` set**: `J: 49.39`, `F: 56.21`, `J&F: 52.80`  

I expected better performance on the val set (`57.0 J&F`), could you clarify why the performance is lower?

### Additional Observations with Smaller Models  

- **Sa2VA-4B**: Works as expected and outputs the `[SEG]` token correctly.  
- **Sa2VA-1B**: Appears to be broken. The output is filled with an unreasonable sequence of exclamation marks:  

```yaml
Model output: !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Request for Clarification  

1. Could you help clarify the descerepancy in reported and observed performance on the MeVIS validation set?  
2. What kind of an issue is there with Sa2VA-1B?

Thank you for your support!

**Comment by zhang-tao-whu** [comment]
Thanks for your reminder. We carefully checked the details and found some errors due to our oversight. The paper reported results on MeVIS val_u set, however this should be corrected to results on val set. We will retest and update our arXiv paper soon. Your evaluation results (val_u J&F 59.01) are basically consistent with our test results (58.9 in Table 8 of the paper). Your finding that uniform sampling achieves better results is something we will try. This makes sense, since the random sampling during training is closer to uniform sampling, and undoubtedly, uniform sampling contains more temporal information compared with first 5 frames sampling.

Additionally, the 1B model might have been corrupted during the conversion process. We will quickly retrain it and update it on HuggingFace tomorrow.

**Comment by lxtGH** [comment]
We have updated 1B model and also release the 26B model for reference.

**Comment by kumuji** [comment]
Hi @lxtGH @zhang-tao-whu ,

I noticed you've updated the code for evaluation and testing; however, I'm having trouble obtaining results from my trained checkpoints. I trained a 4B-parameter model (without sam2 videos) using the provided config, but I can't find a straightforward way to evaluate it or convert it to the HuggingFace format.

Currently, I have a checkpoint in `work_dirs/sa2va_4b/iter_21354.pth/`. The checkpoint loads in the runner just fine.

I attempted to convert the model to hf format so I could use the `projects/llava_sam2/evaluation` scripts, but had no luck. I tried using xtuner’s `pth_to_hf` script, but it calls a `to_hf` method that isn’t in the model definition.

I tried to avoid the HuggingFace format entirely by merging `tools/test.py` and `demo/demo.py`, so I could use a model through `runner.model.module`. This gives me a `VideoLLaVASAMModel_zero3` model, but it lacks the methods defined in `modeling_sa2va_chat.py` from the HuggingFace model, like `predict_forward()`. It seems like it would require quite a bit additional coding to make it work.

Lisa had a script ( [merge_lora_weights_and_save_hf_model.py](https://github.com/dvlab-research/LISA/blob/main/merge_lora_weights_and_save_hf_model.py) ) that merges LoRA weights and saves a HuggingFace model, but you don't provide anything similar.

Could you share a recommended method to evaluate a trained checkpoint? For instance, is there a script or process to directly evaluate checkpoints without converting them to HuggingFace format?

Thank you for your assistance, and help in getting the results!

**Off-topic, but may be useful:**
1. In `ref_vos_val.py`, you import `projects.llava_sam2.hf_evaluation`, but the folder name appears to be `projects.llava_sam2.evaluation`.
2. In `refVOS.py`, there’s an import: `from vlm.datasets.evaluation.base_eval_dataset import BaseEvalDataset` but it looks like `vlm/datasets` doesn’t exist in the repo.

**Comment by zhang-tao-whu** [comment]
@kumuji  Hello, thanks for your attempt. We will provide the conversion script as soon as possible in the coming days.

**Comment by kumuji** [comment]
Hi @zhang-tao-whu @lxtGH ,

It seems that I could reproduce the result for 4B model using uniform frame sampling:
| J&F   | Experiment Name                   |
|-------|-----------------------------------|
| 45.75 |  my training |
| 45.90 | sa2va_4b from hf |

Thank you for your help!

**Comment by HarborYuan** [comment]
Hi @kumuji ,

Thanks for your attempt and sharing the information.
